### PR TITLE
Expose array size limit in IDL where size limit seems intentional

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
We have internal size limits for arrays, and some of them are intentional while some development-restriction (i.e. 254 to say "unlimited").

#### Change overview
Add an array size for attributes in the IDL where the size seems intentional (smaller than 254).

#### Testing
Manual view of the resulting IDL files.

### CODE REVIEW

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter [hunk 1]
```diff
@@ -1823,7 +1823,7 @@ server cluster NetworkCommissioning = 49 {
   }
 
   attribute(readonly) int8u maxNetworks = 0;
-  attribute(readonly) NetworkInfo networks[] = 1;
+  attribute(readonly) NetworkInfo networks[12] = 1;
```
Comment (tcarmelveilleux) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:1826]: Where does the 12 come from?

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter [hunk 2]
```diff
@@ -2180,14 +2180,14 @@ server cluster PowerSource = 47 {
   attribute(readonly) int8u batteryPercentRemaining = 12;
   attribute(readonly) int32u batteryTimeRemaining = 13;
   attribute(readonly) enum8 batteryChargeLevel = 14;
-  attribute(readonly) ENUM8 activeBatteryFaults[] = 18;
+  attribute(readonly) ENUM8 activeBatteryFaults[8] = 18;
```
Comment (tcarmelveilleux) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:2183]: There is a concrete enum for man of these "enum8". The types descend from enum8, but are concrete enums

File: src/app/zap-templates/templates/app/MatterIDL.zapt [hunk 3]
```diff
@@ -42,7 +42,11 @@
     {{~/unless~}}
     ) {{type}} {{asLowerCamelCase name~}}
     {{~#if isList~}}
-      []
+      [
+        {{~#if_valid_list_size_restriction~}}
+          {{maxLength}}
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/MatterIDL.zapt:47]: I don't think we should output this.  It's nonsense for lists and doesn't really get used for anything in Matter except sizing some buffers.... and I plan to stop doing that too.

For _strings_ on the other hand, this value is in fact meaningful (in terms of limiting how long a string we will allow to be written) and we should consider showing it there.

**Review by bzbarsky-apple — DISMISSED** [review]

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter (hunk 2 above)
Comment (bzbarsky-apple) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:2183]: > There is a concrete enum for man of these "enum8". The types descend from enum8, but are concrete enums

In the spec, yes.  The XML we have says:
```
<attribute side="server" code="0x0012" define="POWER_SOURCE_ACTIVE_BAT_FAULTS" type="ARRAY" entryType="ENUM8" length="8" writable="false" optional="true">ActiveBatteryFaults</attribute>
```
and the IDL is faithfully reflecting that....

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter (hunk 1 above)
Comment (andy31415) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:1826]: Umm ... ember ?!? ... probably somewhere in a zap file. This is the reason I am adding this here, to make the "code" (in this case 10K lines of json) reviewable. Will try to find the location.

File: src/app/zap-templates/templates/app/MatterIDL.zapt (hunk 3 above)
Comment (andy31415) [reply src/app/zap-templates/templates/app/MatterIDL.zapt:47]: How do I find if something is a string? Is string the only meaningfully limited item here?

I would like to change if_valid_list_size_restriction to to determine the correct place where sizes are actually sane.

Also I saw no sizes anywhere in "structs" so I assume lists have no explicit bound in zap for structs.

Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/MatterIDL.zapt:47]: Technically, in the matter spec, struct fields can also have caps on size.

In practice, I don't know whether struct fields have a maxLength in ZAP.

For finding whether something is a string, `{{#if (isString type)}}`

File: src/app/zap-templates/partials/idl/structure_member.zapt [hunk 4]
```diff
@@ -1,7 +1,11 @@
 {{~#if isOptional~}} optional {{/if~}}
 {{~#if isNullable~}} nullable {{/if~}}
 
-{{type}} {{asLowerCamelCase label~}}
+{{type}} {{~#if (isString type)~}}
+  {{~#if maxLength~}} 
+    <{{maxLength}}> 
+  {{~/if~}}
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/partials/idl/structure_member.zapt:7]: Just looking at the output, we may be running into the problem where for an array the `type` is the _entry type_...  Gotta love this type system.  So this should probably be:
```suggestion
{{type}} {{~#unless isArray}}{{#if (isString type)~}}
  {{~#if maxLength~}} 
    <{{maxLength}}> 
  {{~/if}}{{/unless~}}
```

File: src/app/zap-templates/templates/app/MatterIDL.zapt [hunk 5]
```diff
@@ -42,7 +42,11 @@
     {{~/unless~}}
     ) {{type}} {{asLowerCamelCase name~}}
     {{~#if isList~}}
-      []
+      [
+        {{~#if_valid_list_size_restriction~}}
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/MatterIDL.zapt:46]: This should still go away, right?

Comment (andy31415) [reply src/app/zap-templates/templates/app/MatterIDL.zapt:46]: good point, forgot. Will remove it.

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter (hunk 1 above)
Comment (andy31415) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:1826]: I tried to find the location ... and failed. If one exists, we can update code later. For now this size will go away soon (forgot to remove it from my current patch version)

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter (hunk 2 above)
Comment (andy31415) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:2183]: Oh yes ... here is where having an IDL starts paying dividents.  Noticing this kind of stuff.

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter (hunk 1 above)
Comment (bzbarsky-apple) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:1826]: The 12 is coming straight from the XML:
```
<attribute side="server" code="0x0001" define="NETWORKS" type="ARRAY" entryType="NetworkInfo" length="12" writable="false" optional="false">Networks</attribute>
```

Good thing we ignore that `length` value.... ;)

File: src/app/zap-templates/partials/idl/structure_member.zapt (hunk 4 above)
Comment (andy31415) [reply src/app/zap-templates/partials/idl/structure_member.zapt]: I did add the check - however with or without it, I saw no output difference.

I am not convinced zap supports "array of strings".

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/zap-templates/templates/app/MatterIDL.zapt [hunk 6]
```diff
@@ -39,9 +39,7 @@
     nosubscribe {{!marker to place a space even with whitespace removal~}}
   {{~/unless~}}
   {{~!Removes spaces~}} attribute {{type}} {{asLowerCamelCase name~}}
-    {{~#if isList~}}
-      []
-    {{~/if}} = {{code}};
+    {{~#if isList~}} [] {{~/if}} = {{code}};
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/MatterIDL.zapt:42]: Shouldn't we include maxLength here for string attributes?  Followup OK for that.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/zap-templates/templates/app/MatterIDL.zapt (hunk 6 above)
Comment (andy31415) [reply src/app/zap-templates/templates/app/MatterIDL.zapt:49]: Done. I keep forgetting that we have two very similar places for data.

**Pull request merged**